src/lib/models/ConfigModel.ts

```diff
@@ -1,172 +1,194 @@
-    #     'severity'          - [Optional] The severity level to apply to this rule by default.
-    #                           Possible values: 1 or 'Critical', 2 or 'High', 3 or 'Moderate', 4 or 'Low', 5 or 'Info'
-    #                           Default: 'Moderate'
-    #     'tags'              - [Optional] The string array of tag values to apply to this rule by default.
-    #                           Default: ['Recommended']
-    #
-    # [Example usage]:
-    # engines:
-    #   regex:
-    #     custom_rules:
-    #       NoTodoComments:
-    #         regex: /\\/\\/[ \\t]*TODO/gi
-    #         file_extensions: [".cls", ".trigger"]
-    #         description: "Prevents TODO comments from being in apex code."
-    #         violation_message: "A comment with a TODO statement was found. Please remove TODO statements from your apex code."
-    #         severity: "Info"
-    #         tags: ['TechDebt']
-    custom_rules: {} # Remove this empty object {} when you are ready to define your first custom rule
-
-  # Custom configuration settings for the 'retire-js' engine
-  # See <LINK_COMING_SOON> to learn more about these settings.
-  retire-js:
-
-    # Whether to turn off the 'retire-js' engine so that it is not included when running Code Analyzer commands
-    disable_engine: false`;
+function getDefaultRule(defaultContext: ConfigContext, engineName: string, ruleName: string): Rule|null {
+	try {
+		return defaultContext.rules.getRule(engineName, ruleName);
+	} catch (e) {
+		// istanbul ignore next
+		return null;
+	}
+}
+
+function toYamlRule(userRule: Rule, defaultRule: Rule|null, styled: boolean): string {
+	const ruleName: string = userRule.getName();
+	const userSeverity: SeverityLevel = userRule.getSeverityLevel();
+	const userTags: string[] = userRule.getTags();
+
+	let severityYaml = yaml.dump({severity: userSeverity});
+	let tagsYaml = yaml.dump({tags: userTags});
+
+	if (defaultRule != null) {
+		const defaultSeverity: SeverityLevel = defaultRule.getSeverityLevel();
+		const defaultTagsJson: string = JSON.stringify(defaultRule.getTags());
+
+		if (userSeverity !== defaultSeverity) {
+			const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [defaultSeverity]);
+			severityYaml = severityYaml.replace('\n', ` ${toYamlComment(comment, styled)}\n`);
+		}
+
+		if (JSON.stringify(userTags) !== defaultTagsJson) {
+			const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [defaultTagsJson]);
+			// The YAML spec requires a trailing newline, so we know that there's at least one newline somewhere in the
+			// string. If we inject a comment before the first newline we encounter, then it will look clean.
+			tagsYaml = tagsYaml.replace('\n', ` ${toYamlComment(comment, styled)}\n`);
+		}
+	}
+	return `"${ruleName}":\n${indent(severityYaml, 2)}${indent(tagsYaml, 2)}`;
+}
+
+function toYamlEngines(relevantEngines: Set<string>, userContext: ConfigContext, styled: boolean): string {
```

Comment:
I believe this function should also receive the defaultContext and then the code would look like:

```
function toYamlEngines(relevantEngines: Set<string>, userContext: ConfigContext, defaultContext: ConfigContext, styled: boolean): string {
	let results: string = 'engines:\n'

	for (const engineName of relevantEngines.keys()) {
		const engineConfigDescriptor = userContext.core.getEngineConfigDescription(engineName);
		const userEngineConfig = userContext.core.getEngineConfig(engineName);
		const defaultEngineConfig = defaultContext.core.getEngineConfig(engineName);

		results += `\n${toYamlComment('='.repeat(70), styled, 2)}\n`
		// Engines are guaranteed to have an overview, even if it's just generic text.
		results += `${toYamlComment(engineConfigDescriptor.overview!, styled, 2)}\n`;
		results += `${toYamlComment('='.repeat(70), styled, 2)}\n`

		results += indent(`${engineName}:`, 2) + '\n';
		// By fiat, the field description will always include, at minimum, an entry for "disable_engine", so we can
		// assume that the object is not undefined.
		for (const configField of Object.keys(engineConfigDescriptor.fieldDescriptions!)) {
			const fieldDescription = engineConfigDescriptor.fieldDescriptions![configField];
			const defaultFieldValue = defaultEngineConfig[configField] ?? null;
			const userFieldValue = userEngineConfig[configField] ?? defaultFieldValue;
			// Add a leading newline to visually break up the property from the previous one.
			results += '\n' + toYamlComment(fieldDescription, styled, 4) + '\n';
			let fieldYaml = indent(`${yaml.dump({[configField]: userFieldValue})}`, 4);
			if(userFieldValue !== defaultFieldValue) {
				const comment = getMessage(BundleName.ConfigModel, 'template.modified-from', [JSON.stringify(defaultFieldValue)]);
				fieldYaml = fieldYaml.replace('\n', `  ${toYamlComment(comment, styled)}\n`);
			}
			results += fieldYaml;
		}
	}
	return results;
}

```